Updated the initcpiocfg module

Boria138 opened the pull request:
Added systemd (I took the code from CachyOS and modified it a bit) 
Fixed the error "setfont: KDFONTOP: Function not implemented"

Boria138 committed:
Updated the initcpiocfg module

Added systemd (I took the code from CachyOS and modified it a bit)
Fixed the error "setfont: KDFONTOP: Function not implemented"

bkmo commented:
Have you also made changes to modify the grub.cfg  GRUB_CMDLINE_LINUX_DEFAULT ?   rd.luks.name= , and also rd.luks.uuid= are needed for when you switch to the systemd hook from udev. But otherwise this is a very welcomed addition.

Boria138 commented:
> Have you also made changes to modify the grub.cfg GRUB_CMDLINE_LINUX_DEFAULT ? rd.luks.name= , and also rd.luks.uuid= are needed for when you switch to the systemd hook from udev. But otherwise this is a very welcomed addition.

Oh, I didn't realize that. I'll correct that mistake later.

bkmo commented:
> > Have you also made changes to modify the grub.cfg GRUB_CMDLINE_LINUX_DEFAULT ? rd.luks.name= , and also rd.luks.uuid= are needed for when you switch to the systemd hook from udev. But otherwise this is a very welcomed addition.
> 
> Oh, I didn't realize that. I'll correct that mistake later.

This is what my grub cmdline looks like:
`rd.luks.name=5d23ccf1-0d39-40c3-b471-8835957583f6=cryptroot root=/dev/mapper/cryptroot rd.luks.uuid=5d23ccf1-0d39-40c3-b471-8835957583f6`

looks to be here in  /src/modules/grubcfg/main.py. The rd.luks additions will need to be added for systemd.

`if partition["mountPoint"] == "/" and has_luks:
                cryptdevice_params = [
                    f"cryptdevice=UUID={partition['luksUuid']}:{partition['luksMapperName']}",
                    f"root=/dev/mapper/{partition['luksMapperName']}"
                ]`

I also just created an issue concerning the file you are basing your commit on. The plymouth section should use encrypt or sd-encrypt  instead  of plymouth-encrypt. That hook is not shipped in the Arch package of Plymouth. Not sure how Manjaro does it though.

Boria138 committed:
Fixes https://github.com/calamares/calamares/issues/2182

Boria138 committed:
Added rd.luks.name to grubcfg

Boria138 commented:
> > > Have you also made changes to modify the grub.cfg GRUB_CMDLINE_LINUX_DEFAULT ? rd.luks.name= , and also rd.luks.uuid= are needed for when you switch to the systemd hook from udev. But otherwise this is a very welcomed addition.
> > 
> > 
> > Oh, I didn't realize that. I'll correct that mistake later.
> 
> This is what my grub cmdline looks like: `rd.luks.name=5d23ccf1-0d39-40c3-b471-8835957583f6=cryptroot root=/dev/mapper/cryptroot rd.luks.uuid=5d23ccf1-0d39-40c3-b471-8835957583f6`
> 
> looks to be here in /src/modules/grubcfg/main.py. The rd.luks additions will need to be added for systemd.
> 
> `if partition["mountPoint"] == "/" and has_luks: cryptdevice_params = [ f"cryptdevice=UUID={partition['luksUuid']}:{partition['luksMapperName']}", f"root=/dev/mapper/{partition['luksMapperName']}" ]`
> 
> I also just created an issue concerning the file you are basing your commit on. The plymouth section should use encrypt or sd-encrypt instead of plymouth-encrypt. That hook is not shipped in the Arch package of Plymouth. Not sure how Manjaro does it though.

Added rd.luks.name to grubcfg, but the sd-encrypt detection is probably not quite right and could have been written better but I mostly write in bash not python

bkmo commented:
I may just fork your repo tomorrow and see how it goes.

dalto8 commented on src/modules/grubcfg/main.py:
```diff
@@ -135,10 +135,13 @@ def modify_grub_default(partitions, root_mount_point, distributor):
     # Shell exit value 0 means success
     have_plymouth = plymouth_bin == 0
     have_dracut = dracut_bin == 0
+    have_systemd = mkinitcpio_systemd == 0 
```
```suggestion
    uses_sd-encrypt = mkinitcpio_systemd == 0 
```

dalto8 commented on src/modules/grubcfg/main.py:
```diff
@@ -214,6 +217,10 @@ def modify_grub_default(partitions, root_mount_point, distributor):
         kernel_params.append(f"resume=/dev/mapper/{swap_outer_mappername}")
+        
+    # When using rd.luks.name you can omit rd.luks.uuid
+    if have_systemd:
+        kernel_params.append(f"rd.luks.name={partition['luksUuid']}={partition['luksMapperName']}")
```
This is going to add this to the existing options even if the partition isn't encrypted.

Also, if the partition is encrypted, it will already have something else added to the options for encryption.

Isn't the logic when using the sd-encrypt the same as when dracut is applied above?  Could we not just use that code?

Boria138 replied:
The mkinitcpio_systemd variable works only if there is sd-encrypt in the hooks, sd-encrypt in turn is added only if unencrypted_separate_boot works in initcpiocfg, that is, in theory, if the partition is not encrypted, it just does not add anything.

dalto8 replied:
What if the distro ships with the sd-encrypt hook by default?

Boria138 replied:
What's supposed to happen? sd-encrypt will not be shipped with an unencrypted partition in any case, so there will be no errors either.

Boria138 replied:
Or do you want to just take the code from dracut and change it somehow ?

```sh
    if uses_sd-encrypt or have_dracut:
        for partition in partitions:
            if partition["fs"] == "linuxswap" and not partition.get("claimed", None):
                # Skip foreign swap
                continue
            has_luks = "luksMapperName" in partition
            if partition["fs"] == "linuxswap" and not has_luks:
                swap_uuid = partition["uuid"]

            if partition["fs"] == "linuxswap" and has_luks:
                swap_outer_uuid = partition["luksUuid"]
                swap_outer_mappername = partition["luksMapperName"]

            if partition["mountPoint"] == "/" and has_luks:
                cryptdevice_params = [f"rd.luks.uuid={partition['luksUuid']}"]
    else:
        for partition in partitions:
            if partition["fs"] == "linuxswap" and not partition.get("claimed", None):
                # Skip foreign swap
                continue
            has_luks = "luksMapperName" in partition
            if partition["fs"] == "linuxswap" and not has_luks:
                swap_uuid = partition["uuid"]

            if partition["fs"] == "linuxswap" and has_luks:
                swap_outer_mappername = partition["luksMapperName"]

            if partition["mountPoint"] == "/" and has_luks:
                cryptdevice_params = [
                    f"cryptdevice=UUID={partition['luksUuid']}:{partition['luksMapperName']}",
                    f"root=/dev/mapper/{partition['luksMapperName']}"
                ]
```

dalto8 replied:
Right, I am asking if the dracut code would work "as-is" when using the sd-encrypt hook.  Assuming we made the code follow that path when the sd-encrypt hook was enabled.

Also, we shouldn't call it `has_systemd`, that would be a very confusing name.  A distro could have systemd and not be using that hook or even mkinitcpio.

Boria138 replied:
I guess if I do as in the snippet above, everything should work correctly

Boria138 replied:
I was originally pointed out the need to redo grubcfg by @bkmo, maybe he can tell me how else to implement it.

dalto8 replied:
If this really requires changes here.  You will probably also need to modify the bootloader module as well.

Boria138 replied:
There's a line in the loader
```sh
if have_dracut:
                cryptdevice_params = [f "rd.luks.uuid={partition['luksUuid']}"]]
```
I guess for systemd it is enough to add this by analogy

```sh
if uses_sd-encrypt:
                cryptdevice_params = [f "rd.luks.name={partition['luksUuid']}={partition['luksMapperName']}"]]
```

dalto8 replied:
You shouldn't need to.  It shouldn't have to be `rd.luks.name=`, `rd.luks.uuid=` should also work.

Boria138 replied:
So you can combine two conditions into one, just like in grubcfg

Boria138 replied:
I seem to have added all the edits that were required of me now waiting for edits on initcpiocfg if they are available

bkmo replied:
does the bootldr module also need an addition for systemd added to the dracut section?

dalto8 reviewed (changes requested):
I haven't reviewed the initcpiocfg changes yet but I have some comments on the grubcfg changes.

Boria138 committed:
Added the necessary edits

vnepogodin commented:
> Added systemd (I took the code from CachyOS and modified it a bit) Fixed the error "setfont: KDFONTOP: Function not implemented"

@Boria138 if you took it from us, then you should probably add us in the license header tbh

Boria138 commented:
> > Added systemd (I took the code from CachyOS and modified it a bit) Fixed the error "setfont: KDFONTOP: Function not implemented"
> 
> @Boria138 if you took it from us, then you should probably add us in the license header tbh

I will add licenses after I pass the review, but can I ask who to write in the license since I found it in the cachyos-calamares-config repository in one of the revent commits and tweaked it a bit I don't remember if anyone was listed in the license there. I will point you and myself in the license as from cachyos I took only when on adding systemd hooks in initcpiocfg all the resulting neanses with encrypted partition I fixed myself so I think so will be quite fair.

Boria138 commented:
I will point you and myself in the license as from cachyos I took only when on adding systemd hooks in initcpiocfg all the resulting neanses with encrypted partition I fixed myself so I think so will be quite fair.

bkmo commented:
I don't see any plagarism from Cachyos. Just inspiration.

Boria138 commented:
> I don't see any plagarism from Cachyos. Just inspiration.

I originally decided to add this after seeing it from cachyos so aside from the binaries and the plymouth part (which had to be modified a bit) the code is theirs so they really should have been added to the license even if they abandoned that code for some reason

bkmo commented:
Ok, i'm sorry, I thought you were working off the main calmares repo, not Cachyos.

Boria138 commented:
By the way about this @vnepogodin if it's not a secret why you finally gave up this code it was partially working until the matter did not concern luks

bkmo commented:
A bit off topic, but what is needed to encrypt with luks2 when boot is unencrypted?

Boria138 commented:
@vnepogodin  I have looked through this revent (https://github.com/CachyOS/calamares-config/commit/b6f16d6df27058ca35ace84723f00a5d2b25367e) from which I took part of the code, I did not find any license or copyright there.

bkmo commented:
I built a package with your branch, but didn't get passed the initcpio stage of the install.

` 'IndentationError' ('unindent does not match any outer indentation level', ('/usr/lib/calamares/modules/initcpiocfg/main.py', 165, 20, '   if uses_systemd:\n', 165, -1))`

Boria138 commented:
> I built a package with your branch, but didn't get passed the initcpio stage of the install.
> 
> ` 'IndentationError' ('unindent does not match any outer indentation level', ('/usr/lib/calamares/modules/initcpiocfg/main.py', 165, 20, '   if uses_systemd:\n', 165, -1))`

Looks like I accidentally made a mistake with the indents tomorrow I'll add a commit that definitely works because I've been testing this with myself

Boria138 committed:
Fixed initcpiocfg

Boria138 commented:
@bkmo I've corrected the indents and a couple of things I inadvertently forgot, I don't have a chance to test the modules yet, but I believe everything should work

bkmo commented:
> @bkmo I've corrected the indents and a couple of things I inadvertently forgot, I don't have a chance to test the modules yet, but I believe everything should work

I'm packaging from your initcpiocfg branch now.

bkmo commented:
I do beleive mkinitcpio was ok, but then this happened:

`'SyntaxError'(&quot;cannot assign to expression here. Maybe you meant '==' instead of '='?&quot;, ('/usr/lib/calamares/modules/grubcfg/main.py', 144, 5, '    uses_sd-encrypt = mkinitcpio_systemd == 0 \n', 144, 20))`

Boria138 committed:
Deleted quot

Boria138 commented:
> I do beleive mkinitcpio was ok, but then this happened:
> 
> `'SyntaxError'(&quot;cannot assign to expression here. Maybe you meant '==' instead of '='?&quot;, ('/usr/lib/calamares/modules/grubcfg/main.py', 144, 5, ' uses_sd-encrypt = mkinitcpio_systemd == 0 \n', 144, 20))`

Apparently the problem was caused by the quotes in grep, so I removed them and that should solve the problem (Although I would appreciate it if someone could tell me how else to track whether a person has an sd-encrypt hook)

bkmo commented:
Still got an error :
` 'SyntaxError'&gt;(&quot;cannot assign to expression here. Maybe you meant '==' instead of '='?&quot;, ('/usr/lib/calamares/modules/grubcfg/main.py', 144, 5, '    uses_sd-encrypt = mkinitcpio_systemd == 0 \n', 144, 20))`

Boria138 commented:
@dalto8 I have repeated the code from the original module, but it gives an error I have no idea what the problem is, maybe you can help me

bkmo commented:
This is the actual error with the html formatting. I think the issue is not with the quotes, but with the == signs. Also what about the empty line 145?

`<div><strong>&lt;class 'SyntaxError'&gt;</strong></div><div>(&quot;cannot assign to expression here. Maybe you meant '==' instead of '='?&quot;, ('/usr/lib/calamares/modules/grubcfg/main.py', 144, 5, '    uses_sd-encrypt = mkinitcpio_systemd == 0 \n', 144, 20))</div>`

Boria138 committed:
uses_sd-encrypt was changed to uses_sd_encrypt to make it a valid variable name

Boria138 commented:
> This is the actual error with the html formatting. I think the issue is not with the quotes, but with the == signs. Also what about the empty line 145?
> 
> `<div>\<strong\>&lt;class 'SyntaxError'&gt;\</strong\></div><div>(&quot;cannot assign to expression here. Maybe you meant '==' instead of '='?&quot;, ('/usr/lib/calamares/modules/grubcfg/main.py', 144, 5, ' uses_sd-encrypt = mkinitcpio_systemd == 0 \n', 144, 20))</div>`

I fixed this

dalto8 commented on src/modules/initcpiocfg/main.py:
```diff
@@ -145,18 +154,29 @@ def find_initcpio_features(partitions, root_mount_point):
+    binaries = []
+
+    # Fixes "setfont: KDFONTOP: Function not implemented" error
+    binaries.append("setfont")
```
What happens if `setfont` isn't installed?  Should we be checking for that condition?

Boria138 replied:
I added check

Boria138 committed:
Added setfont check in mkinitcpiocfg

bkmo commented:
> > This is the actual error with the html formatting. I think the issue is not with the quotes, but with the == signs. Also what about the empty line 145?
> > `<div>\<strong\>&lt;class 'SyntaxError'&gt;\</strong\></div><div>(&quot;cannot assign to expression here. Maybe you meant '==' instead of '='?&quot;, ('/usr/lib/calamares/modules/grubcfg/main.py', 144, 5, ' uses_sd-encrypt = mkinitcpio_systemd == 0 \n', 144, 20))</div>`
> 
> I fixed this

This is not fixed yet , Typo?:

`<div><strong>&lt;class 'NameError'&gt;</strong></div><div>name 'used_sd_encrypt' is not defined</div><div><br/>Traceback:</div><div><pre>File &quot;/usr/lib/calamares/modules/grubcfg/main.py&quot;, line 314, in run
    return modify_grub_default(partitions, root_mount_point, distributor)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^

  File &quot;/usr/lib/calamares/modules/grubcfg/main.py&quot;, line 165, in modify_grub_default
    if have_dracut or used_sd_encrypt:
                      ^^^^^^^^^^^^^^^</pre></div>`

Boria138 committed:
Fixed a stupid typo

Boria138 commented:
> > > This is the actual error with the html formatting. I think the issue is not with the quotes, but with the == signs. Also what about the empty line 145?
> > > `<div>\<strong\>&lt;class 'SyntaxError'&gt;\</strong\></div><div>(&quot;cannot assign to expression here. Maybe you meant '==' instead of '='?&quot;, ('/usr/lib/calamares/modules/grubcfg/main.py', 144, 5, ' uses_sd-encrypt = mkinitcpio_systemd == 0 \n', 144, 20))</div>`
> > 
> > 
> > I fixed this
> 
> This is not fixed yet , Typo?:
> 
> `
> **<class 'NameError'>**
> name 'used_sd_encrypt' is not defined
> 
> Traceback:
> 
> File "/usr/lib/calamares/modules/grubcfg/main.py", line 314, in run
> 
> return modify_grub_default(partitions, root_mount_point, distributor)
> 
> ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
> 
> 
> File "/usr/lib/calamares/modules/grubcfg/main.py", line 165, in modify_grub_default
> if have_dracut or used_sd_encrypt:
> ^^^^^^^^^^^^^^^
> 
> `

It was a very stupid typo, sometimes I'm surprised at my own inattention

bkmo commented:
I forked and fixed it as a test already. I am installing from an Iso in a vm now.

Boria138 commented:
This time I checked the code several times for typos

bkmo commented:
Whwn I saw the keyboard on the password sceen in plymouth I knew it was from the sd-encrypt hook. It booted, used the proper hook. Looks good.

`HOOKS="systemd autodetect kms modconf block keyboard sd-vconsole plymouth sd-encrypt filesystems"`

`GRUB_CMDLINE_LINUX_DEFAULT='quiet rd.luks.uuid=f581310c-60c9-4782-99d7-bc80186b7d9d splash loglevel=3'`

dalto8 commented:
Can either of you link me to an ISO I can use to test this?  All our recent ISOs use dracut.

bkmo commented:
> Can either of you link me to an ISO I can use to test this? All our recent ISOs use dracut.

I don't understand? Do you want an ISO built with the SD calamares commits? Or a config for mkarchiso that you can add it to?

dalto8 commented:
> I don't understand? Do you want an ISO built with the SD calamares commits?

Since these are all python changes, any ISO that has a recent version of Calamares that installs using mkinitcpio should be fine.  I can make the Python changes on the ISO itself before launching Calamares.

bkmo commented:
> > I don't understand? Do you want an ISO built with the SD calamares commits?
> 
> Since these are all python changes, any ISO that has a recent version of Calamares that installs using mkinitcpio should be fine. I can make the Python changes on the ISO itself before launching Calamares.

Ok, let me see what I can find that is pre-built online.  I am thinking ALCI

bkmo commented:
This should do the trick.  https://sourceforge.net/projects/alci/files/alci-iso-lts/archlinux-2023.05.31-x86_64.iso/download

bkmo commented:
> Can either of you link me to an ISO I can use to test this? All our recent ISOs use dracut.

What are you using for Plymouth in your recent ISO's? Are you building it from the current git of Plymouth? Are you adding a plymouth-encrypt hook to the package? Arch is building without the hook and you can use the standard encrypt hook now. The current Calamares still wants to add plymouth-encrypt, but it is non existent on Arch.  Do you think it could be made universal to use just encrypt even though a distro may package with plymouth-encrypt?

dalto8 commented:
In my testing I am seeing `GRUB_ENABLE_CRYPTODISK` being set even with an unencrypted `/boot`.  I wonder if that is a bug in `grubcfg` or an issue with the ALCI ISO.  

@Boria138 I will finish my testing tomorrow and open a PR against your branch with some code quality changes.

bkmo commented:
There is a grub.cfg in the airrootfs of the ALCI iso that has enable cryptodisk uncommented IIRC:
`# Uncomment to enable booting from LUKS encrypted devices
GRUB_ENABLE_CRYPTODISK=y`

Edit: yup it does.  https://github.com/arch-linux-calamares-installer/alci-iso-lts/blob/master/archiso/airootfs/etc/default/grub

bkmo commented:
> @Boria138 I will finish my testing tomorrow and open a PR against your branch with some code quality changes.
>
This is good news as the sd-encrypt hook allows you to open other luks containers with the same password.

bkmo commented:
Why is there a keyfile created, and an entry in cryptab on an unencrypted boot partition?

dalto8 committed:
[initcpiocfg,grubcfg,bootloader] Minor code improvements

dalto8 commented:
@Boria138 I pushed a commit to your branch with some minor changes.  Can you review it and make sure you are OK with it?

@Boria138 , @bkmo can you both test this version when you have a chance?

dalto8 commented:
> Why is there a keyfile created, and an entry in cryptab on an unencrypted boot partition?

Probably because Calamares didn't always have good support for an unencrypted `/boot` and that module has never been updated with a check for it.

bkmo commented:
> @Boria138 , @bkmo can you both test this version when you have a chance?
I'm building it now....

Boria138 commented:
> @Boria138 I pushed a commit to your branch with some minor changes. Can you review it and make sure you are OK with it?

Of course I am only too happy to have your help in improving my code

bkmo commented:
> > Why is there a keyfile created, and an entry in cryptab on an unencrypted boot partition?
> 
> Probably because Calamares didn't always have good support for an unencrypted `/boot` and that module has never been updated with a check for it.

Can you lead me to the modules that would affect this please?

dalto8 commented:
> Can you lead me to the modules that would affect this please?

https://github.com/calamares/calamares/tree/calamares/src/modules/luksbootkeyfile

bkmo commented:
> > Can you lead me to the modules that would affect this please?
> 
> https://github.com/calamares/calamares/tree/calamares/src/modules/luksbootkeyfile

Hmmm....this is in luksbootkeyfilejob.cpp:
`  for ( const auto& d : s.devices )
    {
        // Skip setupLuks for root partition if system has an unencrypted /boot
        if ( d.isRoot && hasUnencryptedSeparateBoot() )
        {
            continue;
        }`

bkmo commented:
The ISO booted up just fine with all the proper settings in grub.fg and mkinitcpio.conf. The setfont binary was also there. Just need to see why a keyfile and cryptab are being created on unencrypted boot partion. At least the keyfile is not added to the initramfs. My builds are configured with luks2 so I did not test with encrypted boot.

dalto8 commented:
> Just need to see why a keyfile and cryptab are being created on unencrypted boot partion.

That is definitely an existing issue.  I noticed the check isn't working in `grubcfg` either.

Perhaps something changed in 3.3 when luks2 support was added.  It needs investigation.

bkmo commented:
> > Just need to see why a keyfile and cryptab are being created on unencrypted boot partion.
> 
> That is definitely an existing issue. I noticed the check isn't working in `grubcfg` either.
> 
> Perhaps something changed in 3.3 when luks2 support was added. It needs investigation.

I'd like to help, but I am not a coder. I am pretty good at reverse engineering things though.

bkmo commented:
> > Just need to see why a keyfile and cryptab are being created on unencrypted boot partion.
> 
> That is definitely an existing issue. I noticed the check isn't working in `grubcfg` either.
> 
> Perhaps something changed in 3.3 when luks2 support was added. It needs investigation.

I was able to stop the keyfile from being generated by adding a different check to luksbootkeyfilejob.cpp:

`
    if ( hasUnencryptedSeparateBoot() && hasEncryptedRoot() )
    {
        // /boot partition is not encrypted, keyfile must not be used
        // But only if root partition is encrypted
        cDebug() << Logger::SubEntry << "/boot partition is not encrypted, skipping keyfile creation.";
        return Calamares::JobResult::ok();
    }`

Now to look into where the cryptab gets created.

dalto8 commented:
That change should probably be part of a separate PR.

bkmo commented:
> That change should probably be part of a separate PR.

I can try to do that.

bkmo commented:
![Screenshot_20230824_182955](https://github.com/calamares/calamares/assets/777373/f9667330-af4e-4bb7-9286-a60476f482ca)
I just built an ISO using Boria138's calamares initcpiocfg branch. I installed arch on a luks1 encrypted root and boot using the erase disk option. It failed to boot. There were issues with sd-vconsole setup, and also finding the root device. A reinstall with an unencrypted boot installed fine.

Boria138 commented:
> 
> ![Screenshot_20230824_182955](https://github.com/calamares/calamares/assets/777373/f9667330-af4e-4bb7-9286-a60476f482ca)
> I just built an ISO using Boria138's calamares initcpiocfg branch. I installed arch on a luks1 encrypted root and boot using the erase disk option. It failed to boot. There were issues with sd-vconsole setup, and also finding the root device. A reinstall with an unencrypted boot installed fine.

About console fonts it is quite expected behavior because calamares does not write fonts in the config, I wrote a separate shell process for that, but about boot it needs more tests

Boria138 commented:
I guess you still need a separate code for sd-encrypt and not combine it with dracut

bkmo commented:
The strange part is that it works fine with a separate unencrypted boot partition. No boot issues, or consolefont issues. I tried editing the grub commandline without a result. On my daily driver I have systemd working fine. I also have an rd.luks.name and a root= commands in the kernel commandline of both grub and systemd-boot. IIRC I needed both when I set it up. I will look at it one more time later as I have a feeling it lies in the kernel commandline UUID's or something like that.

bkmo commented:
There has to be something different with grub when using an encrypted vs non encrypted /boot. The only difference with the encrypted /boot failing vs not failing is that grub is handling the unlocking and not sd-encrypt. Maybe the cryptdevice is still needed in grub cmndline. Beats me. But it does work fine in an unencrypted /boot.

bkmo commented:
Ok, after grub opens the luks container to load the initramfs, it then needs an encrypt hook to open it again. This is what is missing. SD-encrypt is not added to hooks. The second issue is there is now another password promt. What will be needed is to figure out what sd-encrypt needs to unlock via a keyfile. This is done with Crptdevice= with the encrypt hook.

`if encrypt_hook:
        if unencrypted_separate_boot:
            if uses_systemd:
                hooks.append("sd-encrypt")
            else:
                hooks.append("encrypt")`

I prpose removing `if unencrypted_seperate_boot:`

Boria138 commented:
@dalto8 review this code snippet please

```sh
if partition["mountPoint"] == "/" and has_luks:
    cryptdevice_params = [
        f"cryptdevice=UUID={partition['luksUuid']}:{partition['luksMapperName']}",
        f"root=/dev/mapper/{partition['luksMapperName']}"
    ]
    crypto_file = "crypto_keyfile.bin"
    if os.path.isfile(
        os.path.join(root_mount_point, crypto_file)
    ):
        cryptdevice_params = [
            f"cryptdevice=UUID={partition['luksUuid']}:{partition['luksMapperName']}",
            f"root=/dev/mapper/{partition['luksMapperName']}",
            f"cryptkey=UUID={partition['luksUuid']}:{partition['fs']}:/{crypto_file}"
        ]
 ```
This for add keyfile to grub

Boria138 commented:
This for add keyfile to grub

dalto8 committed:
[initcpiocfg] Fix encryption hook not being added with encrypted /boot

dalto8 commented:
> I prpose removing `if unencrypted_seperate_boot:`

Yeah, that was definitely wrong.  I pushed a fix for that.



> @dalto8 review this code snippet please

Hmm....I am not sure about that.  Let me do some testing.

bkmo commented:
> 

I'm building a pkg now based on Boria138's initcpio branch. All that needs to be done is getting sd-encrypt to use a keyfile from grub.

The package installed and booted fine. There was a second password prompt as expected.

Boria138 commented:
> > 
> 
> I'm building a pkg now based on Boria138's initcpio branch. All that needs to be done is getting sd-encrypt to use a keyfile from grub.

I have written adding a key file when using sd-encrypt I just need to check if this code works, and for that we need to wait for dalto

dalto8 committed:
[grubcfg] Add rd.luks.key for systemd-encrypt hook

dalto8 commented:
I just pushed https://github.com/calamares/calamares/pull/2183/commits/3552691e57eed06b7ef939c6e5bd5d72d972919e which I think resolves that issue.

bkmo commented:
> I just pushed [3552691](https://github.com/calamares/calamares/commit/3552691e57eed06b7ef939c6e5bd5d72d972919e) which I think resolves that issue.

I will see how it goes.

bkmo commented:
That's all she wrote. Booted up just fine with only the initial grub password prompt. We have both encrypted and unencrypted boot working with systemd hooks. I will double check unencrypted install and report back here.

Boria138 commented:
@dalto8 since all the tests are passed, I guess we can merge the tree.

bkmo commented:
I'm double checking an unencrypted install, but I assume it will be fine.

bkmo commented:
Unencrypted install pases just fine, plymouth and all. Now we can revisit the lukskey PR. It works too but there is awarning.

ptr1337 commented:
As already discussed with dalto:
There should be a config toggle or a detection for which filesystem this does apply.
This should be currently NOT enabled as default. The systemd hook breaks:
1. ZFS Installations (any bootloader)
2. Rescue shell (https://bugs.archlinux.org/task/70408)
3. fsck and other scripts are not anymore detected and used anymore

Maybe there is something else on top, which I dont have seen so far.
The systemd hook does only bring a improvement at the boot time - are there any others?

I don't think we should break the above topics and existing features in calamares for a faster bootime.

bkmo commented:
I prefer systemd sd-encrypt over the encrypt hook because it will unlock multiple luks containers with the same password. Encrypt will only open the one container. The boot time savings is negligible iIMHO. So for me it is a nice addition, but if it  breaks things maybe a config option would work. As far as rescue shell goes, adding in the base hook should cure that.
according to the Arch WIKI:
This will install a basic systemd setup in your initramfs, and is meant to replace the 'base', 'usr', 'udev' and 'timestamp' hooks. Other hooks with runtime components will need to be ported, and will not work as intended. You also may wish to still include the 'base' hook (before this hook) to ensure that a rescue shell exists on your initramfs.

Boria138 commented:
The best option would be to add the systemd setting to the module's config file, but I don't have the knowledge for such a thing

dalto8 commented:
I will add that option.

bkmo commented:
> I will add that option.
And if the rescue shell is a concern you can add the base hook before the systemd hook in mkinitcpio.conf. What harm can it do? Just a bit larger initramfs.

dalto8 commented:
> And if the rescue shell is a concern you can add the base hook before the systemd hook in mkinitcpio.conf. What harm can it do? Just a bit larger initramfs.

It doesn't help.  It just tells you the root account is locked.

bkmo commented:
> > And if the rescue shell is a concern you can add the base hook before the systemd hook in mkinitcpio.conf. What harm can it do? Just a bit larger initramfs.
> 
> It doesn't help. It just tells you the root account is locked. 

Good to know. I have never tried it. It must be missing a systemd target.

ptr1337 commented:
> > And if the rescue shell is a concern you can add the base hook before the systemd hook in mkinitcpio.conf. What harm can it do? Just a bit larger initramfs.
> 
> It doesn't help. It just tells you the root account is locked.

Only replacing the "udev" hook helps in that mkinitcpio won't complain anymore about the missing scripts.

bkmo commented:
I'm happy with no base and no udev for systemd. I can't remember if I have ever used the rescue shell. I have always fixed things in chroot. Maybe just add a systemd option to initcpio.conf

ptr1337 commented:
I just made some further testing:
Encrypted Installation (BTRFS, XFS, F2FS,..) are broken when systemd as bootloader is used.
Also when the recent introduced config variable is set.

In my eyes, this should be reverted till it is correctly fixed and working.
@bkmo 
Did you not wanted this PR, because of lvm encrypted installation? Did you test this together with systemd-boot?

dalto8 commented:
The error you received has been fixed

Boria138 commented:
@dalto8 In 6.5 setfont doesn't work, so the check for it should be removed otherwise the error when loading, I know that I added it, but before the error was without it, and now with it.

Boria138 commented:
setfont: ERROR kdfontop.c:183 put_font_kdfontop: Unable to load such font with such kernel version `

This is the boot error if I use binaries maybe the problem is in the kernel from cachyos I will then put zen kernel and check it out

dalto8 commented:
Reverted: https://github.com/calamares/calamares/commit/f3f5bd8a5d2a6dc9cc1f0560542c195628d0b6d1

Honestly, I was never fully comfortable with that part anyway.  It seemed like a workaround so I think it is best to revert it for now.

Boria138 commented:
@ptr1337 I'm using your calamares version instead of upstream since the release is too old and the git version may have critical bugs and it's too hard to maintain my branch, so could you please synchronize your fork with upstream so that the kernel bug doesn't happen, but most importantly add only initcpiocfg fixes, since the move to qt6 seems to have broken translations.

vnepogodin commented:
> @ptr1337 I'm using your calamares version instead of upstream since the release is too old and the git version may have critical bugs and it's too hard to maintain my branch, so could you please synchronize your fork with upstream so that the kernel bug doesn't happen, but most importantly add only initcpiocfg fixes, since the move to qt6 seems to have broken translations.

we use up-to-date version. the repo from which you got is very old and we dont use it since year or so

Boria138 commented:
> > @ptr1337 I'm using your calamares version instead of upstream since the release is too old and the git version may have critical bugs and it's too hard to maintain my branch, so could you please synchronize your fork with upstream so that the kernel bug doesn't happen, but most importantly add only initcpiocfg fixes, since the move to qt6 seems to have broken translations.
> 
> we use up-to-date version. the repo from which you got is very old and we dont use it since year or so
> 

For my distribution I use the standard branch from here
https://github.com/CachyOS/cachyos-calamares
Since I don't need cachyos customizations and custom modules just a more stable version of calamares, and I took some of the code for adding systemd hooks from cachyos calamares config because it was there and the new version didn't have it.

Boria138 commented:
Your version of calamares was updated a week ago, so you have the bug with setfont, well, and another bug because of which it is not possible to use systemd hooks but it is not critical, but because of the bug with setfont I have on your kernel stopped starting the system.

Boria138 commented:
Kinda like the setfont bug doesn't happen if you use systemd hooks, but I haven't tested it.

dalto8 commented:
PR comments are not the right place for this discussion.  Please move this conversation to a more appropriate location.

kkofler commented:
It is not a good idea to depend on another distro's Calamares fork without even discussing that with them beforehand, all the more if you are unable to maintain a fork yourself.
